File: Python/sorting/merge_sort.py
def mergesort(arr):
    """
    MergeSort algorithm implementation
    Time Complexity: O(n log n) in all cases
    Space Complexity: O(n) for temporary arrays
    """
    # Base case: arrays with 0 or 1 element are already sorted
    if len(arr) <= 1:
        return arr
    
    # Divide: find the middle point
    mid = len(arr) // 2
    
    # Recursively sort left and right halves
    left_half = mergesort(arr[:mid])
    right_half = mergesort(arr[mid:])
    
    # Conquer: merge the sorted halves
    return merge(left_half, right_half)


def merge(left, right):
    """
    Merge two sorted arrays into one sorted array
    """
    result = []
    i = j = 0
    
    # Compare elements from left and right, add smaller to result
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    
    # Add remaining elements (one of these will be empty)
    result.extend(left[i:])
    result.extend(right[j:])
    
    return result


def mergesort_inplace(arr, left=0, right=None):
    """
    In-place MergeSort implementation (optimized for space)
    Modifies the original array
    """
    if right is None:
        right = len(arr) - 1
    
    if left < right:
        # Find the middle point
        mid = (left + right) // 2
        
        # Recursively sort first and second halves
        mergesort_inplace(arr, left, mid)
        mergesort_inplace(arr, mid + 1, right)
        
        # Merge the sorted halves
        merge_inplace(arr, left, mid, right)
    
    return arr


def merge_inplace(arr, left, mid, right):
    """
    Merge helper function for in-place mergesort
    Merges two sorted subarrays arr[left..mid] and arr[mid+1..right]
    """
    # Create temporary arrays
    left_arr = arr[left:mid + 1]
    right_arr = arr[mid + 1:right + 1]
    
    i = j = 0
    k = left
    
    # Merge temporary arrays back into arr[left..right]
    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1
    
    # Copy remaining elements
    while i < len(left_arr):
        arr[k] = left_arr[i]
        i += 1
        k += 1
    
    while j < len(right_arr):
        arr[k] = right_arr[j]
        j += 1
        k += 1


# Test the implementations
if __name__ == "__main__":
    # Test data
    test_arrays = [
        [64, 34, 25, 12, 22, 11, 90],
        [5, 2, 8, 1, 9],
        [1],
        [],
        [3, 3, 3, 3],
        [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
    ]
    
    print("=" * 50)
    print("MERGESORT - Simple Version")
    print("=" * 50)
    
    for arr in test_arrays:
        original = arr.copy()
        sorted_arr = mergesort(arr)
        print(f"Original: {original}")
        print(f"Sorted:   {sorted_arr}\n")
    
    print("=" * 50)
    print("MERGESORT - In-Place Version")
    print("=" * 50)
    
    for arr in test_arrays:
        original = arr.copy()
        mergesort_inplace(arr)
        print(f"Original: {original}")
        print(f"Sorted:   {arr}\n")
    
    # Performance test with larger array
    import random
    import time
    
    large_array = [random.randint(1, 1000) for _ in range(1000)]
    
    start = time.time()
    sorted_array = mergesort(large_array.copy())
    end = time.time()
    print(f"Simple MergeSort - 1000 elements: {(end - start) * 1000:.2f}ms")
    
    start = time.time()
    mergesort_inplace(large_array.copy())
    end = time.time()
    print(f"In-place MergeSort - 1000 elements: {(end - start) * 1000:.2f}ms")
    
    # Interactive input (uncomment to use)
    # print("\n" + "=" * 50)
    # print("INTERACTIVE MODE")
    # print("=" * 50)
    # n = int(input("Enter number of elements: "))
    # arr = list(map(int, input("Enter elements separated by space: ").split()))
    # sorted_arr = mergesort(arr)
    # print("Sorted array:", *sorted_arr)

File: Dart/sorting/merge_sort.dart
/// Merge sort implementation in Dart
/// Time Complexity: O(n log n)
/// Space Complexity: O(n)

/// Merge sort implementation
List<T> mergeSort<T extends Comparable<T>>(List<T> arr) {
  if (arr.length <= 1) return List<T>.from(arr);
  
  int mid = arr.length ~/ 2;
  List<T> left = arr.sublist(0, mid);
  List<T> right = arr.sublist(mid);
  
  return merge(mergeSort(left), mergeSort(right));
}

/// Merge two sorted lists into one sorted list
List<T> merge<T extends Comparable<T>>(List<T> left, List<T> right) {
  List<T> result = [];
  int i = 0, j = 0;
  
  while (i < left.length && j < right.length) {
    if (left[i].compareTo(right[j]) <= 0) {
      result.add(left[i]);
      i++;
    } else {
      result.add(right[j]);
      j++;
    }
  }
  
  // Add remaining elements
  while (i < left.length) {
    result.add(left[i]);
    i++;
  }
  
  while (j < right.length) {
    result.add(right[j]);
    j++;
  }
  
  return result;
}

/// In-place merge sort implementation (modifies original array)
void mergeSortInPlace(List<int> arr, [int? left, int? right]) {
  left ??= 0;
  right ??= arr.length - 1;
  
  if (left < right) {
    int mid = left + (right - left) ~/ 2;
    
    mergeSortInPlace(arr, left, mid);
    mergeSortInPlace(arr, mid + 1, right);
    
    _mergeInPlace(arr, left, mid, right);
  }
}

/// Helper function for in-place merge
void _mergeInPlace(List<int> arr, int left, int mid, int right) {
  // Create temporary arrays
  List<int> leftArr = arr.sublist(left, mid + 1);
  List<int> rightArr = arr.sublist(mid + 1, right + 1);
  
  int i = 0, j = 0, k = left;
  
  // Merge the temporary arrays back into arr[left..right]
  while (i < leftArr.length && j < rightArr.length) {
    if (leftArr[i] <= rightArr[j]) {
      arr[k] = leftArr[i];
      i++;
    } else {
      arr[k] = rightArr[j];
      j++;
    }
    k++;
  }
  
  // Copy remaining elements
  while (i < leftArr.length) {
    arr[k] = leftArr[i];
    i++;
    k++;
  }
  
  while (j < rightArr.length) {
    arr[k] = rightArr[j];
    j++;
    k++;
  }
}

/// Bottom-up merge sort implementation
List<T> mergeSortBottomUp<T extends Comparable<T>>(List<T> arr) {
  if (arr.length <= 1) return List<T>.from(arr);
  
  List<T> result = List<T>.from(arr);
  int n = result.length;
  
  // Merge subarrays in bottom-up manner
  for (int size = 1; size < n; size *= 2) {
    for (int left = 0; left < n - size; left += 2 * size) {
      int mid = left + size - 1;
      int right = (left + 2 * size - 1).clamp(0, n - 1);
      
      if (mid < right) {
        List<T> leftPart = result.sublist(left, mid + 1);
        List<T> rightPart = result.sublist(mid + 1, right + 1);
        List<T> merged = merge(leftPart, rightPart);
        
        for (int i = 0; i < merged.length; i++) {
          result[left + i] = merged[i];
        }
      }
    }
  }
  
  return result;
}

/// Merge sort with custom comparator
List<T> mergeSortWithComparator<T>(List<T> arr, int Function(T, T) compare) {
  if (arr.length <= 1) return List<T>.from(arr);
  
  int mid = arr.length ~/ 2;
  List<T> left = arr.sublist(0, mid);
  List<T> right = arr.sublist(mid);
  
  return _mergeWithComparator(
    mergeSortWithComparator(left, compare),
    mergeSortWithComparator(right, compare),
    compare
  );
}

List<T> _mergeWithComparator<T>(List<T> left, List<T> right, int Function(T, T) compare) {
  List<T> result = [];
  int i = 0, j = 0;
  
  while (i < left.length && j < right.length) {
    if (compare(left[i], right[j]) <= 0) {
      result.add(left[i]);
      i++;
    } else {
      result.add(right[j]);
      j++;
    }
  }
  
  result.addAll(left.sublist(i));
  result.addAll(right.sublist(j));
  
  return result;
}

void main() {
  List<int> unsortedList = [38, 27, 43, 3, 9, 82, 10];
  print('Merge Sort Examples:');
  print('Original: $unsortedList');
  print('Sorted (functional): ${mergeSort(unsortedList)}');
  
  // Test in-place merge sort
  List<int> inPlaceList = List<int>.from(unsortedList);
  mergeSortInPlace(inPlaceList);
  print('Sorted (in-place): $inPlaceList');
  
  print('Sorted (bottom-up): ${mergeSortBottomUp(unsortedList)}');
  
  // Test with strings
  List<String> words = ['merge', 'sort', 'algorithm', 'divide', 'conquer'];
  print('\nSorting words: $words');
  print('Result: ${mergeSort(words)}');
  
  // Test with custom comparator (descending order)
  List<int> descendingSort = mergeSortWithComparator(unsortedList, (a, b) => b.compareTo(a));
  print('\nDescending sort: $descendingSort');
}